# CART Implementaion Using Scikit-Learn Library

## Decision Tree
Decision Trees are commonly used in data mining with the objective of creating a model that predicts the value of a target (or dependent variable) based on the values of several input (or independent variables). 

## CART
The CART or Classification & Regression Trees methodology was introduced in 1984 by Leo Breiman, Jerome Friedman, Richard Olshen and Charles Stone as an umbrella term to refer to the following types of decision trees:

### Classification Tree
Where the target variable is categorical and the tree is used to identify the "class" within which a target variable would likely fall into.
### Regression Trees
Where the target variable is continuous and tree is used to predict it's value.

The main elements of CART (and any decision tree algorithm) are:
* Rules for splitting data at a node based on the value of one variable;
* Stopping rules for deciding when a branch is terminal and can be split no more; and
* Finally, a prediction for the target variable in each terminal node.

### CART Model Representation
The representation for the CART model is a binary tree.

This is your binary tree from algorithms and data structures, nothing too fancy. Each root node represents a single input variable and a split point on that variable (assuming the variable is numeric).

The leaf nodes of the tree contain an output variable which is used to make a prediction.

#### Advantages
* CART is nonparametric and therefore does not rely on data belonging to a particular type of distribution.
* CART is not significantly impacted by outliers in the input variables.
* You can relax stopping rules to "overgrow" decision trees and then prune back the tree to the optimal size.  This approach minimizes the probability that important structure in the data set will be overlooked by stopping too soon.
* CART incorporates both testing with a test data set and cross-validation to assess the goodness of fit more accurately.
* CART can use the same variables more than once in different parts of the tree.  This capability can uncover complex interdependencies between sets of variables.
* CART can be used in conjunction with other prediction methods to select the input set of variables.

#### Disadvantages
* Overfitting
* Decision tree can be unstable because of small change in data.
* It cannot guarantee to return globally optimal decision tree.
* Biased Trees

### Loading the modules required, from ScikitLearn Library:

In [2]:
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier, export_graphviz, DecisionTreeRegressor
from sklearn.model_selection import train_test_split

from sklearn.externals.joblib import Memory
from sklearn.datasets import load_svmlight_file

c:\users\amrit\anaconda3\envs\myenv\lib\site-packages\sklearn\externals\joblib\__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=FutureWarning)


### Loading Other modules

In [6]:
import graphviz 
import csv
import pandas as pd 
import numpy as np
import time

### Enabling Inline Plotting

In [7]:
%matplotlib inline

### Loading Dataset

In [8]:
# Load a CSV file
def load_data(filename, filetype):
    if filetype == 'csv':
        dataset = pd.read_csv(filename)
        nrow, ncol = dataset.shape
        X = dataset.iloc[:, :ncol-1]
        Y = dataset.iloc[:, ncol-1:ncol]
        return X, Y
    
    elif filetype == 'libsvm':
        mem = Memory("./mycache")
        @mem.cache
        def get_data():
            data = load_svmlight_file(filename)
            return data[0], data[1]
        X, Y = get_data()
        return X, Y
    
    else:
        print('File Type Not Supported !')

### Prepare data 

#### Glass Identification Data Set 
From USA Forensic Science Service; 6 types of glass; defined in terms of their oxide content (i.e. Na, Fe, K, etc).
Number of Instances : 214
Number of Attributes : 10
Attribute Information:

1. Id number: 1 to 214 
2. RI: refractive index 
3. Na: Sodium (unit measurement: weight percent in corresponding oxide, as are attributes 4-10) 
4. Mg: Magnesium 
5. Al: Aluminum 
6. Si: Silicon 
7. K: Potassium 
8. Ca: Calcium 
9. Ba: Barium 
10. Fe: Iron 
11. Type of glass: (class attribute) 
    1. building_windows_float_processed 
    2. building_windows_non_float_processed 
    3. vehicle_windows_float_processed 
    4. vehicle_windows_non_float_processed (none in this database) 
    5. containers 
    6. tableware 
    7. headlamps



In [9]:
# Dataset Splitting For Training and Testing
X, Y = load_data('glass', 'libsvm')
data = X
target = Y

data_train, data_test, target_train, target_test = train_test_split(data, target, test_size=0.33, random_state=42)

________________________________________________________________________________
[Memory] Calling __main__-E%3A-DM_Git-CART-__ipython-input__.get_data...
get_data()
_________________________________________________________get_data - 0.0s, 0.0min


In [10]:
# fit a CART model to the data
cfTree = DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=10,
            max_features=None, max_leaf_nodes=10,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

regTree = DecisionTreeRegressor(random_state=0)
regTree.fit(data_train, target_train)

startTime = time.time()
cfTree.fit(data_train, target_train)
endTime = time.time()
timeDiff = endTime - startTime
print(timeDiff)

0.0039980411529541016


c:\users\amrit\anaconda3\envs\myenv\lib\site-packages\sklearn\tree\_classes.py:319: FutureWarning: The parameter 'presort' is deprecated and has no effect. It will be removed in v0.24. You can suppress this warning by not passing any value to the 'presort' parameter.
  FutureWarning)


In [11]:
# make predictions
expectedClass = target_test
predictedClass = cfTree.predict(data_test)

In [12]:
# summarize the fit of the model
print(metrics.classification_report(expectedClass, predictedClass))

              precision    recall  f1-score   support

         1.0       0.70      0.73      0.71        22
         2.0       0.71      0.68      0.69        25
         3.0       0.25      0.25      0.25         4
         5.0       0.67      0.67      0.67         6
         6.0       0.75      0.75      0.75         4
         7.0       1.00      1.00      1.00        10

    accuracy                           0.72        71
   macro avg       0.68      0.68      0.68        71
weighted avg       0.72      0.72      0.72        71



In [13]:
print(metrics.confusion_matrix(expectedClass, predictedClass))

[[16  4  2  0  0  0]
 [ 5 17  1  2  0  0]
 [ 2  1  1  0  0  0]
 [ 0  1  0  4  1  0]
 [ 0  1  0  0  3  0]
 [ 0  0  0  0  0 10]]


In [14]:
# Calculate accuracy percentage
def accuracy_metric(actual, predicted):
    correct = 0
    for i in range(len(actual)):
        if actual[i] == predicted[i]:
            correct += 1
    return correct / float(len(actual)) * 100.0

In [15]:
accuracy = accuracy_metric(expectedClass, predictedClass)
print(accuracy)

71.83098591549296


In [ ]:
# Export the CF tree
dot_data = export_graphviz(cfTree, out_file=None) 
graph = graphviz.Source(dot_data) 
graph.render("DecisionTreeClassifier")

# Export the Reg tree
dot_data = export_graphviz(regTree, out_file=None) 
graph = graphviz.Source(dot_data) 
graph.render("DecisionTreeRegression")

## Pruning
<table>
    <tr>
        <th>Split Criterion</th>
        <th>Max Depth</th>
        <th>Max Leaf Node</th>
        <th>Run time(in s)</th>
        <th>Accuracy(%)</th>
    </tr>
    <tr>
        <td>Gini</td>
        <td>10</td>
        <td>10</td>
        <td>0.0043</td>
        <td>74.64</td>
    </tr>
    <tr>
        <td>Gini</td>
        <td>10</td>
        <td>5</td>
        <td>0.0040</td>
        <td>70.42</td>
    </tr>  
    <tr>
        <td>Gini</td>
        <td>5</td>
        <td>10</td>
        <td>0.0029</td>
        <td>70.42</td>
    </tr>
    <tr>
        <td>Gini</td>
        <td>5</td>
        <td>5</td>
        <td>0.0033</td>
        <td>71.83</td>
    </tr> 
    <tr>
        <td>Entropy</td>
        <td>5</td>
        <td>10</td>
        <td>0.0050</td>
        <td>61.97</td>
    </tr> 
    <tr>
        <td>Entropy</td>
        <td>10</td>
        <td>10</td>
        <td>0.0054</td>
        <td>61.97</td>
    </tr>
    <tr>
        <td>Entropy</td>
        <td>5</td>
        <td>5</td>
        <td>0.0051</td>
        <td>59.15</td>
    </tr>
    <tr>
        <td>Entropy</td>
        <td>10</td>
        <td>5</td>
        <td>0.0061</td>
        <td>59.15</td>
    </tr>
</table>

### Conclusion
* From the above we can deduce that Gini split is giving better accuracy than Entropy, So Gini Index can be the better option for the split criterion. And the runtime for Gini Split is also very less as compared to entropy.
* In all the cases if we are increasing the depth of the tree the accuracy is increasing.
